### ISSUE TITLE
Causes firefox to hang while loading

### ISSUE THREAD

**Issue by shadowmint** [webcomponents/webcomponentsjs#75]
In firefox the terminal logs:

    mutating the [[Prototype]] of an object will cause your code to run very slowly; instead create the    
    object with the correct initial [[Prototype]] value using Object.create

As webcomponents.js loads it causes the browser to temporarily hang even when the tab is not active.

Performance log

![perf](http://i.imgur.com/kkzYCfv.png)

NB. This is a carry over issue from platform.js that was a performance regression from 0.4.0 that never got addressed.

**Comment by deepaksisodiya** [comment]
I guess, webcomponents is a platform independent library, You must solve this kind of problem. I am using requirejs with webcomponents and because of this performance issue, I am getting timeout error from requirejs.
I rquest you to solve this bug and show love for Firefox ;)

**Comment by neshkatrapati** [comment]
+1

**Comment by jmesserly** [comment]
fyi, mutating the prototype is unavoidable, and has been there for as long as the polyfill has existed, so is unlikely to be related. Once Firefox implements Custom Elements and Shadow DOM natively, performance will be much better. Until then, you may have luck with webcomponents-lite.js, which doesn't include the (very expensive) Shadow DOM polyfill.

**Comment by Mikanoshi** [comment]
I hope FF will have native support soon, because current polyfills are extremely slow. Even IE is faster...
https://github.com/Polymer/polymer/issues/952

**Comment by shadowmint** [comment]
@garlicnation why have you closed this? It is *still* unresolved.

**Comment by jmesserly** [comment]
@shadowmint -- if you can provide more concrete info about the problem, we can take a look. I'm not sure what action to take from just looking at the perf screenshot (it looks like minified, so it doesn't provide any information like stack trace). If the culprit is ShadowDOM polyfill, there isn't much we can do until Firefox has native support. (If you're using Polymer, they are hard at work on the 0.8 branch, which will at least not depend so heavily on SD polyfill)

**Comment by shadowmint** [comment]
more concrete info?

Firefox *hangs* all tabs when you open a page with a component on it using
this polyfill.

What would you like me to provide? A codepen example? I can do that if it
will help.

**Comment by garlicnation** [comment]
I closed the wrong issue when working on a related one. When debugging another SD bug my laptop was pegged at 100% cpu for the multi-second pageload time.

I think this is caused by changes in firefox, not the SD polyfill, but I'm investigating further to determine a root cause.

@shadowmint a codepen example would be great!

**Comment by shadowmint** [comment]
Here's an example pen: http://codepen.io/shadowmint/embed/wCceK/

It may seem to open 'ok' in firefox, but watch what happens when you hit reload. The entire browser actually hangs during the load; opening firebug makes this even more obvious, but firebug always makes everything +50% slower anyway.

As the number of component instances increases, you'll experience the joy of bitbucket, where one of your tabs which is open entirely destroys the browser.

Just cut and paste a bunch of `<hello-header id="copy2"></hello-header>` into the html section of the pen to see it happen live. 

fwiw notice both that this is using 0.5.3, not 0.5.4, but there's no meaningful difference in terms of performance, and that if you use webcomponents-lite.js this isn't an issue, but that's because using lite doesn't have a shadowdom, so the component code fails to load entirely.

ie. tldr; Could well be the shadowdom polyfil on firefox causing the issue, but you can't actually use components without it, so.... /shrug

**Comment by Mikanoshi** [comment]
This is how you hang FF :)
http://codepen.io/Mikanoshi/pen/jEGVOj

ShadowDOM polyfill is just too slow, same code in browsers with native WebComponents support take 1-2 seconds.

**Comment by jmesserly** [comment]
@Mikanoshi yup, that one is unfortunately not fixable though. If using SD polyfill, the best idea is to avoid creating a huge DOM. Something like https://github.com/Polymer/core-list which can load nodes on demand.

I did a lot of research on a [faster SD polyfill](https://github.com/jmesserly/webcomponentsjs/tree/faster_shadowdom), and it looks like it might be possible in Firefox/IE, but unfortunately that approach can't work on Safari because of a [bug](https://bugs.webkit.org/show_bug.cgi?id=49739). So you'd need a special minifier, and most people didn't like that idea.

**Comment by Mikanoshi** [comment]
IE is not a huge loss :) Safari? Who needs Safari. I'll just add a disclaimer that FF is slow with this site.
I have ~500 paper-inputs divided into 15 core-pages, they are loaded dynamically, so only 20-40 polymer elements are on page at the same time, but performance is still bad.

Loading time is not the biggest problem, polymer elements are doing something in a loop, that makes everything on the page stutter, it's very noticeable during transitions (try slider in my codepen example).

**Comment by jmesserly** [comment]
@Mikanoshi fyi Polymer 0.8-preview only depends on webcomponents-lite.js so that will hopefully make your experience better in the future :)

**Comment by Mikanoshi** [comment]
@jmesserly Lite doesn't have ShadowDOM, but all polymer elements are based on templates, so it will be a very distant future :)

**Comment by garlicnation** [comment]
Confirmed that this behavior is the same with 0.5.3 and 0.5.4 and is the same in past and current firefoxes.

With that said...

The ShadowDom polyfill will definitely be unnecessary with 0.8, and it will be clear how to upgrade existing elements.

**Comment by Mikanoshi** [comment]
Is FF team not going to add native support? What are they waiting?

**Comment by jmesserly** [comment]
they are working on it https://hacks.mozilla.org/2014/12/mozilla-and-web-components/
bug tree for web components https://bugzilla.mozilla.org/showdependencytree.cgi?id=811542

you can actually go into about:config in firefox and enable "dom.webcomponents.enabled" flag. That makes APIs like createShadowRoot work.

**Comment by Mikanoshi** [comment]
That flag breaks all polymer sites completely.

**Comment by ben-8409** [comment]
I think i am affected by this as well on Firefox. Interestingly, it does only happen on my own example running locally – if i open some of the polymer examples, like http://1.daily-stock-forecast.appspot.com/, it is way faster (100times faster). I only have one element on my page, can not using "vulcanize" really be the problem here?

**Comment by shadowmint** [comment]
Just idly, this continue to break sites using polymer like https://events.google.com/io2015/ in firefox.

It's probably worth updating the 'browser support' on https://github.com/webcomponents/webcomponentsjs for firefox to '~' rather than '✓' for shadow dom.

**Comment by schipplock** [comment]
I get this warning even with "webcomponents-lite". The polymer site says "Production ready" though. I feel betrayed :). Google style. Meeh.

**Comment by jmesserly** [comment]
> I get this warning even with "webcomponents-lite". The polymer site says "Production ready" though. I feel betrayed :). Google style. Meeh.

it's from CustomElements, which is part of -lite:
https://github.com/webcomponents/webcomponentsjs/blob/master/src/CustomElements/upgrade.js#L75

There's nothing slow about that line of code, it's exactly how everyone working on CustomElements is doing polyfills. Mozilla folks are using -lite in their own projects. Firefox's warning message is highly misleading, for this particular case. I have nothing against the message, most code mutating `__proto__` is not doing it in a safe way, but we are in this case. (Technical reason: because it happens immediately after the object is born, and only the immediate `__proto__` of that object, not something further up the chain.)

Whatever the problem is here, I'm 100% sure it has nothing to do with the `__proto__` warning.

**Comment by MagmaRules** [comment]
I'm also seeing this behavior. Firefox hangs while loading the page.

**Comment by TimvdLippe** [comment]
This is currently blocking us from deploying as the FireFox page load hangs for up to 8 seconds before starting to show content. In retrospect, the same page in Chrome loads in < 150 ms. @garlicnation any update or hint we can use to mitigate this?

**Comment by samccone** [comment]
@TimvdLippe can you outline the steps to reproduce? Thanks!

**Comment by TimvdLippe** [comment]
@samccone So I tried to reproduce this, but it is very hard actually ;p

One thing that became apparent (later I read @MagmaRules discovered the same) when opening the inspector/debugging tab the page is very slow. Without the console tools, the page loads just a little bit slower than Chrome, but this was expected as it does not run native.

1. With the Inspector tab open, the loading time roughly doubles.
* With the Debugger tab, it is roughly four times slower.
* With the Console tab open, the page does not want to load at all. I had to open a different tab to unblock and continue loading.

So I assume this is an issue on Firefox side, but I honestly have no clue what the root issue could be. I have tested this with a moderately sized project with the Polymer Starter Kit architecture.

Concluding, I think this issue can be closed. It of course makes debugging and developing with Firefox very hard/impossible, but I am not sure if that is something for the Polymer team to fix?

**Comment by samccone** [comment]
@TimvdLippe would you be able to put together a very simple test case? I can pull in some people from the firefox side that can help us here once we have a good reproducible test case :)

**Comment by shadowmint** [comment]
I dont understand... when did this issue become about firefox with the
developer tools open?

I issue I reported was firefox freezing.

The polyfill is still unusable for firefox for the same reason, as far as
I'm aware.

If a different issue is now somehow being tracked here, please create a new
ticket for it, and leave this one open.

...unless the polyfill *is* fixed now?

If so, what fixed the performance regression I initally reported?

What version of the polyfill are you using with it working that I can test
against?

**Comment by samccone** [comment]
@shadowmint would you be able to put together a test case?

**Comment by TimvdLippe** [comment]
@shadowmint when testing this issue, I could not reproduce it with the latest version of webcomponents. Therefore I commented I assumed this issue was solved on the way down. That said, I will create a new issue for the tools issue in the next few days once I am able to get a small reproducible test case.

**Comment by shadowmint** [comment]
There already is one in the comment history.

**Comment by shadowmint** [comment]
Right.

I've done a bunch of testing on firefox on various platforms and basically I'm happy to let this issue die.

Here is a pen that appears to work correctly using firefox 44.0.1 and webcomponent.js 0.7.20:

http://codepen.io/anon/pen/PZxjXb

I don't see the slow down I initially reported, and the cross-tab browser killing freeze appears to have gone as well. 

\o/

So, I'm going to close this issue.

Feel free to link to a new issue here if you create a new one to track developer tools issue / etc.

**Comment by kao98** [comment]
> Feel free to link to a new issue here if you create a new one to track developer tools issue / etc.

Done: #493 opened ;)

I think the issue is is due to Firefox developer tools, but as samccone said, he

> can pull-in some people from the firefox side

so I hope the new issue will be helpfull !?

**Comment by TimvdLippe** [comment]
Oops I completely forgot to create the issue. Thanks @kao98 !
